Refactoring All JS-Scripts

Brandon-T opened the pull request:
## Security Review
- https://github.com/brave/security/issues/1039

## Summary of Changes
- Open for initial draft review only
- Refactor almost scripts to be self-contained
- All dynamic scripts have their own unique security token
- All static scripts have the same security token they've always had
- Message handlers are done automatically unless otherwise specified (all done within `TabContentScript`)
- ContentWorlds are handled within the script itself, and not within a script manager
- Refactor all scripts to not expose tokens or functions that use tokens
- Override toString on all postMessage functions so they cannot be printed.
- Freeze message handlers so they cannot be changed by the page.
- Fix: https://github.com/brave/internal/issues/903

TODO:
- Figure out how to freeze handlers for static scripts (scripts that are automatically injected)
- Get rid of BrowserViewController deciding what to inject, and have each tab decide itself based on preferences

<!-- Enter a ticket number for this PR, create a new one if it is not there yet. -->
This pull request fixes N/A

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:
<!-- Any useful notes explaining how best to test and verify. -->
Test plan in https://github.com/brave/internal/issues/903

## Screenshots:
<!-- If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here. -->


## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue and pull request is assigned to a milestone (should happen at merge time).

Brandon-T committed:
Optimize scripts

Brandon-T committed:
Formatting + more beefing up

Brandon-T committed:
Freezing types

Brandon-T committed:
Finished `__firefox__.js` script rewrite.
Everything is secured.

iccub reviewed (commented):
1. I would double check if the `TabContentScript` properties have to be static
2. Would be good to create a small markdown document and maybe explain how to create new scripts in a secure way

Other than that code looks clean to me, I skipped testing more complex js parts as it's too much for me. Will do a functional review shortly

Brandon-T committed:
Securing scripts from fake webkit and freezing prototype of UserScriptMessageHandler

Brandon-T committed:
Fixes injecting handlers

Brandon-T committed:
Sandboxed DeAmp

Brandon-T committed:
Make a more secure version of `webit.messageHandlers.handlerName.PostMessage`.
Refactors!
All scripts working again?

Brandon-T committed:
Add other parameters

Brandon-T committed:
Fixed sandbox handlers

Brandon-T committed:
Bind Call, Apply, Bind recursively.
Even if pages override these, it will not create a stack-overflow.
Even if pages override these, and freeze the prototype, it won't affect protected functions and scripts.
Benefits: We can freeze these functions while still keeping the Prototype functions not frozen!

Brandon-T committed:
Fixed object prototype freezing.

iccub reviewed (changes requested):
1. Please double check all scripts are inserted into proper main or non-frame contexts. One of scripts I added has been readded by you in non-main context so it's worth checking others too IMO
2. Fix tests

Rest looks solid to me

iccub committed:
Correct name for SiteStateListener script

Brandon-T committed:
Fixed frame for BraveSkus

Brandon-T committed:
Address comments & Fixed Tests

Brandon-T committed:
Use shadow element to hide cosmetic filters

Brandon-T committed:
Fixed sandbox scripts manual evaluation

Brandon-T committed:
Remove logging

thypon reviewed (commented):
I'm still not completely sure if I understand correctly. Do we still need message handlers' security tokens, or are they now not needed anymore since we may delete the handler from being accessible from the page altogether?

I want to give it a second look tomorrow since the change touches everything.

Brandon-T committed:
Remove obfuscating handler for sandboxed script

stoletheminerals reviewed (approved):
I suggest doing complete functional testing. I reported bugs I encountered, but I didn't test every script-related functionality. Other than that lgtm!

Brandon-T committed:
Add BraveSkusScript which was missing due to gitignore.

Brandon-T committed:
Protecting all scripts

Brandon-T committed:
Refactor all script injection and handlers

Brandon-T committed:
Update scripts: BraveSkus, and SiteStateListener

Brandon-T committed:
Addressed feedback 1.

Brandon-T committed:
Removed __firefox__.js
